### PR TITLE
Add `native_dynamic_linking` to `test_data`

### PR DESCRIPTION
**Pull request by blaugold**
This PR adds test data for https://github.com/dart-lang/sdk/issues/56459.

CL that adds tests in Dart SDK that use the test_data: https://dart-review.googlesource.com/c/sdk/+/381580

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/blob/main/docs/External-Package-Maintenance.md#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

File: pkgs/native_assets_builder/test_data/native_dynamic_linking/hook/build.dart [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+      await builder.run(
+        config: config,
+        output: output,
+        logger: logger,
+      );
+    }
+  });
+}
+
+extension on BuildConfig {
```
Comment (mosuem) [inline pkgs/native_assets_builder/test_data/native_dynamic_linking/hook/build.dart:57]: Code like this, switching over OSs, seems like something which should actually live in a helper package. Or is it specific to this example?

Comment (mosuem) [reply pkgs/native_assets_builder/test_data/native_dynamic_linking/hook/build.dart:57]: Or is this covered by https://github.com/dart-lang/native/issues/190?

**Review by dcharkes — COMMENTED** [review]
Please exercise the test data in the tests:

* run the hook
* check if there's more than one asset
* try to dlopen (probably requires spawning a new Dart process with the right path).

After that, LGTM! 🚀

File: pkgs/native_assets_builder/test/test_data/native_dynamic_linking_test.dart [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+      final processResult = await Process.run(
+        dartUri.toFilePath(),
+        [
+          'hook/build.dart',
+          '-Dout_dir=${tempUri.toFilePath()}',
+          '-Dpackage_name=$name',
+          '-Dpackage_root=${testPackageUri.toFilePath()}',
+          '-Dtarget_os=${OSImpl.current}',
+          '-Dversion=${HookConfigImpl.latestVersion}',
+          '-Dlink_mode_preference=dynamic',
```
Comment (dcharkes) [inline pkgs/native_assets_builder/test/test_data/native_dynamic_linking_test.dart:47]: add `-Dlinking_enabled=0` (see fixes to tests in https://github.com/dart-lang/native/pull/1507)

Or possibly better, use `BuildConfig` constructor and write the config to a file and pass that in.

**Review by dcharkes — APPROVED** [review]

File: pkgs/native_assets_builder/test/test_data/native_dynamic_linking_test.dart (hunk 2 above)
Comment (blaugold) [reply pkgs/native_assets_builder/test/test_data/native_dynamic_linking_test.dart]: Done

**Review by dcharkes — APPROVED** [review]

**Pull request merged**